### PR TITLE
content tag

### PR DESCRIPTION
**Pull request by patricklx**
I did the update for eslint, and though that it might work for this as well.

design:
1. Run `Preprocessor.parse` from content-tag to get template locations. manually convert gjs files with `<template>` tags into valid JS, which doesn't change location. The `<template>` tag gets converted into empty objects.
2. convert empty object AST node to a custom `Template` Ast node, which holds the template content.
3. Run the `estree` Prettier printer, which formats the valid JS above.
4. Grab template contents from `Template` AST node described above.
5. Run the hbs Prettier printer against the template contents.
6. Replace the `Template` AST node with the results from above, wrapped in `<template>`

others:
`</template> as TemplateOnlyComponent<Signature>` will be left to prettiers default handling

### CODE REVIEW

**Comment by patricklx** [comment]
@gitKrystan i think this works. But you would need to remove most of the ambiguous cases...

**Comment by gitKrystan** [comment]
Thanks. I will look into this ASAP. Might not be until next week though.

**Comment by patricklx** [comment]
btw, i didn't remove the tests for the ambiguous parts. Wanted to wait on content-tag.

**Comment by NullVoxPopuli** [comment]
> Wanted to wait on content-tag.

how do you mean? is there an issue / failing test for that project?

**Comment by gitKrystan** [comment]
Can you re-run `pnpm install`? Seems like the lock file is out of date.

**Comment by patricklx** [comment]
https://github.com/embroider-build/content-tag/issues/37

**Comment by gitKrystan** [comment]
The purpose of the ambiguous cases in the tests is that we don't introduce syntax errors when prettifying.

e.g.

```ts
<template>Hello World</template>
(oops) => {}
```

We should be able to detect this case and print:

```ts
<template>Hello World</template>;
(oops) => {}
```

**Comment by NullVoxPopuli** [comment]
I'm adding failing tests for those to content-tag now :tada:

**Comment by patricklx** [comment]
> The purpose of the ambiguous cases in the tests is that we don't introduce syntax errors when prettifying.
> 
> e.g.
> 
> ```ts
> <template>Hello World</template>
> (oops) => {}
> ```
> 
> We should be able to detect this case and print:
> 
> ```ts
> <template>Hello World</template>;
> (oops) => {}
> ```

How would it happen that you introduce syntax errors?
Maybe it cannot happen anymore now?

**Comment by gitKrystan** [comment]
I have lots of thoughts, but no idea which ones are relevant. Please push the lockfile changes so I can run the tests and see what you are talking about. 🙏

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -29,7 +29,7 @@ jobs:
 
       - run: pnpm install --frozen-lockfile
 
-      - run: pnpm lint
+      # - run: pnpm lint
```
Comment (patricklx) [inline .github/workflows/ci.yml:32]: Will fix the lint issues later

Comment (NullVoxPopuli) [reply .github/workflows/ci.yml:32]: who not pnpm lint --fix?

**Comment by gitKrystan** [comment]
Looks like there are a few different types of test failure:

### Missing visitor keys for 'Template'.

Cause unknown. This seems to be the majority of the failures.

### Ambiguous cases that are now syntax errors.

With the previous `ember-template-imports` parsing, these were not a syntax error.

Example of a more realistic case.

```ts
const ListItem = <template><li>My Item</li></template> // no semi

<template>
  <ul>
    <ListItem />
    <ListItem />
    <ListItem />
  </ul>
</template>
```

If `content-tag` team says these are now syntax errors, we can continue to error as we do now and remove the relevant tests. (See below first, though.)

If `content-tag` team says these are not syntax errors, the issue needs to be fixed in the `content-tag` repo.

### Change of behavior in other ambiguous cases.

These are still not considered syntax errors in the `content-tag` world, but how we are handling it seems to have changed. Will have to diff the snapshot output to determine how, but it's probably some combo of new-lines and semi-colons.

Here's an example of where the behavior change appears to be introducing a syntax error where there was none. The user added a semi-colon to prevent the syntax error and this plugin removed it.

```
[test:run]  FAIL  tests/unit-tests/ambiguous/arrow-parens-avoid.test.ts > ambiguous > config > arrowParens: "avoid" > (oh, no) => {} > with semi, with newline > it formats ../cases/gjs/default-export.gjs
[test:run] Error: Snapshot `ambiguous > config > arrowParens: "avoid" > (oh, no) => {} > with semi, with newline > it formats ../cases/gjs/default-export.gjs 1` mismatched
[test:run] 
[test:run] - Expected
[test:run] + Received
[test:run] 
[test:run]   "<template>
[test:run]     Explicit default export module top level component. Explicit default export
[test:run]     module top level component. Explicit default export module top level
[test:run]     component. Explicit default export module top level component. Explicit
[test:run]     default export module top level component.
[test:run] - </template>;
[test:run] + </template>
[test:run]   (oh, no) => {};
[test:run]   "
[test:run] 
[test:run]  ❯ behavesLikeFormattedAmbiguousCase tests/helpers/ambiguous.ts:95:20
[test:run]      93|   try {
[test:run]      94|     const result = await format(code, formatOptions);
[test:run]      95|     expect(result).toMatchSnapshot();
[test:run]        |                    ^
[test:run]      96|   } catch (error: unknown) {
[test:run]      97|     // Some of the ambiguous cases are Syntax Errors when parsed
[test:run]  ❯ tests/helpers/ambiguous.ts:64:13
```

### Change of behavior with config `templateExportDefault: true`

Cause unknown. Maybe this case just isn't handled yet. With `templateExportDefault: true`, we should include `export default` for the relevant template export.

```
[test:run]  FAIL  tests/unit-tests/config/template-export-default.test.ts > config > templateExportDefault: true > it formats ../cases/gjs/default-export.gjs
[test:run] Error: Snapshot `config > templateExportDefault: true > it formats ../cases/gjs/default-export.gjs 1` mismatched
[test:run] 
[test:run] - Expected
[test:run] + Received
[test:run] 
[test:run] - "export default <template>
[test:run] + "<template>
[test:run]     Explicit default export module top level component. Explicit default export
[test:run]     module top level component. Explicit default export module top level
[test:run]     component. Explicit default export module top level component. Explicit
[test:run]     default export module top level component.
[test:run]   </template>
[test:run]   "
[test:run] 
[test:run]  ❯ tests/helpers/make-suite.ts:59:20
[test:run]      57|     const code = testCase.code.replaceAll(AMBIGUOUS_PLACEHOLDER, '');
[test:run]      58|     const result = await format(code, config.options);
[test:run]      59|     expect(result).toMatchSnapshot();
[test:run]        |                    ^
[test:run]      60|   });
[test:run]      61| }
```

**Comment by patricklx** [comment]
I think i missed something when pushing...
Locally i didn't have `Missing visitor keys for 'Template'.` and had like ~300 failing tests
Will check tomorrow.
Basically there were only syntax errors.

**Comment by gitKrystan** [comment]
Thanks again for working on this!

**Comment by patricklx** [comment]
mmm, still 600 failing tests in ci... I'll check again tomorrow

**Comment by patricklx** [comment]
vitest was somehow auto updating all tests snapshots... this will need more work

**Comment by patricklx** [comment]
okay, now down to 360

File: tests/unit-tests/ambiguous/__snapshots__/arrow-parens-avoid.test.ts.snap [hunk 2]
```diff
@@ -726,7 +726,7 @@ oops => {};
 exports[`ambiguous > config > arrowParens: "avoid" > (oops) => {} > with semi, with newline > it formats ../cases/gjs/simple.gjs 1`] = `
 "<template>
   what
-</template>
+</template>;
```
Comment (patricklx) [inline tests/unit-tests/ambiguous/__snapshots__/arrow-parens-avoid.test.ts.snap:729]: I think this is how prettier would handle it if template is transformed to function invocation

**Comment by patricklx** [comment]
I can probably fix the semi change by changing the ast type to function expression for the template. I think.

**Comment by NullVoxPopuli** [comment]
After reverting all the snapshot changes, it looks like a _lot_ of the failures are unexpected `export default`

**Comment by patricklx** [comment]
Yes. There is also 1 bug at
https://github.com/gitKrystan/prettier-plugin-ember-template-tag/pull/162/files#diff-ed9c6e22579caeb649360bf79dac90d5222c8de99f5c31968cf53ff3825175eeL379

But note that i only add the export default if it's enabled in the options:
https://github.com/gitKrystan/prettier-plugin-ember-template-tag/pull/162/files#diff-fdd9b7cec43f9e7c596cd66d5e1656a3fc2ad2c0ae98cc7a165273a560a022e1R105

File: tests/unit-tests/__snapshots__/format.test.ts.snap [hunk 3]
```diff
@@ -381,7 +383,7 @@ export interface Signature {
   module top level component. Explicit default export module top level
   component. Explicit default export module top level component. Explicit
   default export module top level component.
-</template> as TemplateOnlyComponent<Signature>
+</template> as TemplateOnlyComponent<Signature>;
```
Comment (patricklx) [inline tests/unit-tests/__snapshots__/format.test.ts.snap:386]: according to prettier and having <template> expressed as an functionDeclaration AST node this should have a semi
@gitKrystan ?

File: tests/unit-tests/ambiguous/__snapshots__/arrow-parens-avoid.test.ts.snap [hunk 4]
```diff
@@ -6,7 +6,7 @@ exports[`ambiguous > config > arrowParens: "avoid" > (oh, no) => {} > with semi,
   module top level component. Explicit default export module top level
   component. Explicit default export module top level component. Explicit
   default export module top level component.
-</template>;
```
Comment (patricklx) [inline tests/unit-tests/ambiguous/__snapshots__/arrow-parens-avoid.test.ts.snap:9]:
```js
function template() {}
(oh, no) => {};
```
is without semi
https://prettier.io/playground/#N4Igxg9gdgLgprEAuEAzArlMMCW0AE8AtgA4A2AhvABQCU+wAvgDpTUQAWANPlBPQF4AfA0YBuViC4gIJXNADOyUBQBOqiAHcACmoRKUFMpooBPJdIBGqimADWcGAGUKROABkcUOMlRGFcFY29o5OJLZeAObIMKrogSABRDgxcQlwAB4kcKo4brBGACo5UGo4cAZ+ZAHSClFkcACK6BDwvv4JAFYKGU71TS1tSFU1IACOg3DaGiQGIBQKALTecAAma1IgsRQ4ZFEAwhBERBTI82Rkm3VQkQ0AgjCxOJbo8No5nt7t1QkcMERkADqHBw8AU4TAcCc+lBOAAbqDTGcwAoLCA4fEAJJQdawJxgXJyO44pwwUwNb6jEgaAKAmwkM7Uio5OE+aReAKqGDTCiRE6UhLhVScs5k7IKAk4OSbaleGCAnCrGAcZAADgADNJVHAJjhtTy+adhh1pDAKJYFUqVUgAEzSdABQrmyomkBwIiWNbrVbuCg3dC8uAAMQgqhOjyiZworwgIEYjCAA

**Comment by patricklx** [comment]
@gitKrystan are you okay with this changes to semi handling?
note that in the the end < template > will be converted to a function call, not a function declaration.
shall we change the ast node to `CallExpression` instead?

**Comment by patricklx** [comment]
now all green

**Comment by patricklx** [comment]
nope, example is failing. content-tag does not like vite build.
can we just tsc build to dist, or do you want to have this work in browser env as well?

File: tests/helpers/ambiguous.ts [hunk 5]
```diff
@@ -96,7 +96,9 @@ async function behavesLikeFormattedAmbiguousCase(
   } catch (error: unknown) {
     // Some of the ambiguous cases are Syntax Errors when parsed
     const isSyntaxError =
-      error instanceof SyntaxError || String(error).startsWith('SyntaxError');
+      error instanceof SyntaxError ||
+      String(error).startsWith('SyntaxError') ||
+      String(error).startsWith('Error: Parse Error');
```
Comment (patricklx) [inline tests/helpers/ambiguous.ts:101]: This was missing to catch parsing errors from content-tag

**Comment by NullVoxPopuli** [comment]
> content-tag does not like vite build.

Content-tag in browsers won't work until this is merged: https://github.com/embroider-build/content-tag/pull/18

cc @ef4

File: tests/unit-tests/ambiguous/__snapshots__/arrow-parens-avoid.test.ts.snap [hunk 6]
```diff
@@ -986,7 +986,7 @@ oops => {};
 exports[`ambiguous > config > arrowParens: "avoid" > (oops) => {} > with semi, without newline > it formats ../cases/gjs/simple.gjs 1`] = `
 "<template>
   what
-</template>
+</template>;
 oops => {};
```
Comment (patricklx) [inline tests/unit-tests/ambiguous/__snapshots__/arrow-parens-avoid.test.ts.snap:990]: I'm not sure why this happens... Looking into it

Comment (patricklx) [reply tests/unit-tests/ambiguous/__snapshots__/arrow-parens-avoid.test.ts.snap:990]: yes, somehow it ended up inside an ExpressionStatement...

File: tests/unit-tests/ambiguous/__snapshots__/arrow-parens-avoid.test.ts.snap [hunk 7]
```diff
@@ -68,7 +68,7 @@ const bool = false,
 exports[`ambiguous > config > arrowParens: "avoid" > (oh, no) => {} > with semi, with newline > it formats ../cases/gjs/simple.gjs 1`] = `
 "<template>
   what
-</template>;
+</template>
 (oh, no) => {};
```
Comment (patricklx) [inline tests/unit-tests/ambiguous/__snapshots__/arrow-parens-avoid.test.ts.snap:72]: this will not cause syntax errors anymore, as content-tag uses a real ast for transforms instead of just string replacements

**Comment by patricklx** [comment]
@gitKrystan I think this is ready.
I wonder if the ambiguous handling is still relevant with content-tag?
its using a custom parser to transform the code, so syntax errors will cause an error here as well. 
So now the  ambiguous handling only has impact at runtime. maybe it  should be the responsibility of eslint or glint to warn the user of constructs like 
`<template></template> + x`
`<template></template>['as']`

File: tests/unit-tests/ambiguous/__snapshots__/arrow-parens-avoid.test.ts.snap (hunk 7 above)
Comment (patricklx) [reply tests/unit-tests/ambiguous/__snapshots__/arrow-parens-avoid.test.ts.snap:72]: wrong, it will cause syntax errors...

**Comment by patricklx** [comment]
@gitKrystan i was able to minimize the changes to semi handling.
what do you think?
i also thought of another way to check if a semi is required or not. by printing the next node with semi: false, look at the result and then fixup the previous builders.doc result.
i will cleanup the code more later.

File: src/print/index.ts [hunk 8]
```diff
@@ -67,66 +81,118 @@ export const printer: Printer<Node | undefined> = {
...
+        args,
+      );
+      const flat = flattenDoc(printedSemiFalse);
+      const previousTemplatePrinted = path.node.extra[
+        'prevTemplatePrinted'
+      ] as builders.Doc;
+      const previousFlat = flattenDoc(previousTemplatePrinted);
+      if (flat[0]?.startsWith(';') && previousFlat.at(-1) !== ';') {
+        if (Array.isArray(previousTemplatePrinted)) {
+          previousTemplatePrinted.push(';');
```
Comment (patricklx) [inline src/print/index.ts:113]: I will simply this part, by wrapping the previous print in an array

File: tests/unit-tests/__snapshots__/format.test.ts.snap (hunk 3 above)
Comment (gitKrystan) [reply tests/unit-tests/__snapshots__/format.test.ts.snap:386]: Seems fine.

File: tests/unit-tests/__snapshots__/format.test.ts.snap [hunk 9]
```diff
@@ -119,8 +119,10 @@ exports[`format > config > default > it formats ../cases/gjs/exported-mod-var.gj
 
 exports[`format > config > default > it formats ../cases/gjs/invalid-template.gjs 1`] = `
 "<template>
+  
               {{#if oops}}
-            </template>
+
```
Comment (gitKrystan) [inline tests/unit-tests/__snapshots__/format.test.ts.snap:124]: Are we adding newlines on syntax errors now? This was a bug in the past.

File: tests/unit-tests/ambiguous/__snapshots__/index.test.ts.snap [hunk 10]
```diff
@@ -1938,17 +1916,16 @@ export interface Signature {
   Yields: [];
 }
 
-<template>
+export default (<template>
```
Comment (gitKrystan) [inline tests/unit-tests/ambiguous/__snapshots__/index.test.ts.snap:1919]: Looks like we picked up an accidental `export default` here.

File: tests/unit-tests/ambiguous/__snapshots__/arrow-parens-avoid.test.ts.snap (hunk 6 above)
Comment (gitKrystan) [reply tests/unit-tests/ambiguous/__snapshots__/arrow-parens-avoid.test.ts.snap]: Probably because of the ambiguity. 🙃

**Comment by gitKrystan** [comment]
Thanks again for all this work!

Re: the cases causing syntax errors. Let's keep them for now since they get swallowed anyway. As we've discovered, what counts as a syntax error might change as the implementation of content-tag evolves.

Re: ambiguous cases, we want to follow Prettier's current patterns as discussed [here](https://github.com/gitKrystan/prettier-plugin-ember-template-tag/issues/1). We can/should update that analysis if necessary since content-tag doesn't produce an array like `ember-template-imports` did. [Check out this example.](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEB6VACArgZzgJwwBMBLIjKCGDAQyPLwFsTIAbaAOgwggAccMAczgwBAHRA5WWfLzhEJ3WBAwwAFnAw4ajTaxJQ4Y2HEa9WNeAAoAlMYDaAch79HAXWPH02PIVLkDMGkiTSYWCHYoLl58ERgSAgwAdxJWVgwAazg4XgwSGAAaDDgANwQ8qAoIAFowjEYIEK9MK3UILEE1POoUtIwAI00GsvIYFXVNQcEDKANBbgAzVQ0KOAAPan1DCqra0xJ6xrg7EzMLaxsAbgdnPhx3EAKQPnjoHGRQGnx8CCSABS+CHeKBorCSNAAnu8nv18DQwFkYABlHRwAAyBjgyAWoLwMLhCJESN48LmyBg+CwcCeYXJlOpIHWcnwJF0sFBABUCFAvglgTjWHjJHNWHAAIpYKhYpACoUAKxwayRIvFkvg2NxDIAjmq4H8fvxkCAaDhqoZ5PJHiAKTRUnMAMIQRiMGhG0GsK04FUAQRgFJI-Sw8D+BAxhg1goZahgjFYAHU1Pk4DgSWA4EigfkSCV8hCjWAcNCQCUqQBJKAhWBIsAs3gwb0VpEwCGiiNCmIQPBxuG8I0xZMEMpWgy+GD6miCF1thkk-C+I39GiDVjVURWmIGGBxsjqZAADgADE9YjqSLFx5PXTLNU8YEvt0Rd0gAExPXBwDlL-k3xmMQb0eQ0RoKBBCwCc4AAMQgfAXT9MkQSDCAQAAX2QoA)

File: tests/unit-tests/ambiguous/__snapshots__/arrow-parens-avoid.test.ts.snap (hunk 6 above)
Comment (patricklx) [reply tests/unit-tests/ambiguous/__snapshots__/arrow-parens-avoid.test.ts.snap]: Right, i fixed it

File: tests/unit-tests/ambiguous/__snapshots__/index.test.ts.snap (hunk 10 above)
Comment (patricklx) [reply tests/unit-tests/ambiguous/__snapshots__/index.test.ts.snap:1919]: its because i'm leaving  `as TemplateOnlyComponent<Signature>` to prettiers default handling.
Since it now remains an expression, we cannot safely remove or add default exports to it.
are you okay with this?

File: tests/unit-tests/__snapshots__/format.test.ts.snap (hunk 9 above)
Comment (patricklx) [reply tests/unit-tests/__snapshots__/format.test.ts.snap]: fixed it. how about now?
```hbs
<template>
  {{#if oops}}
</template>
```

**Comment by patricklx** [comment]
ambiguous cases:
i think the only difference is when there are multiple templates in succession. then no semi is needed anymore.
```hbs
<template></template>
<template></template>
```
its either wrapped in a `static {}` for class properties or exported default. or will end up with a function call, which also works without semi

File: tests/unit-tests/ambiguous/__snapshots__/index.test.ts.snap (hunk 10 above)
Comment (gitKrystan) [reply tests/unit-tests/ambiguous/__snapshots__/index.test.ts.snap:1919]: Seems fine.

File: examples/input/example.gts [hunk 11]
```diff
@@ -1,12 +1,14 @@
 import type { TemplateOnlyComponent } from '@ember/component/template-only';
 
 export interface Signature {
-  Element: HTMLElement,
-  Args: {
-
-
-  }
-  Yields: []
+  Element: HTMLElement;
```
Comment (gitKrystan) [inline examples/input/example.gts:4]: I think we can revert the changes in this file.

File: tests/unit-tests/__snapshots__/format.test.ts.snap (hunk 9 above)
Comment (gitKrystan) [reply tests/unit-tests/__snapshots__/format.test.ts.snap]: This is better, but if our goal is to follow Prettier's existing behavior, it seems that syntax errors should no-op.

Comment (patricklx) [reply tests/unit-tests/__snapshots__/format.test.ts.snap]: Oh, okay then

Comment (patricklx) [reply tests/unit-tests/__snapshots__/format.test.ts.snap]: fixed

File: examples/input/example.gts (hunk 11 above)
Comment (patricklx) [reply examples/input/example.gts]: fixed

File: src/parse.ts [hunk 12]
```diff
@@ -1,220 +1,243 @@
...
-      isGlimmerClassProperty(node)
-    ) {
-      getGlimmerExpression(node).extra.isDefaultTemplate = true;
-    }
-  };
+interface PreprocessedResult {
+  templateVisitorKeys: Record<string, string[]>;
+  templateInfos: {
+    templateRange: [number, number];
+    range: [number, number];
```
[the remark this answers is not in the document]
Comment (patricklx) [reply src/parse.ts]: added 
```js
/**
     * range of the template including <template></template> tags
     */
    templateRange: [number, number];
    /**
     * range of the template content, excluding <template></template> tags
     */
    range: [number, number];
```

**Comment by patricklx** [comment]
@gitKrystan squashed commits, its ready now

**Review by gitKrystan — APPROVED** [review]
